### PR TITLE
fix gasUsed and fee for relayed v1 and v2

### PR DESCRIPTION
**Pull request by sstanculeanu**
## Reasoning behind the pull request
- node api does not take into account the latest fix for base cost of relayed when computing fees with refund
  
## Proposed changes
- proper fix

## Testing procedure
- system test + scenarios

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: outport/process/transactionsfee/transactionsFeeProcessor.go [hunk 1]
```diff
@@ -249,6 +271,19 @@ func (tep *transactionsFeeProcessor) prepareTxWithResultsBasedOnLogs(
 
 	for _, event := range txWithResults.log.GetLogEvents() {
 		if core.WriteLogIdentifier == string(event.GetIdentifier()) && !hasRefund {
+			if !check.IfNilReflect(userTx) && tep.enableEpochsHandler.IsFlagEnabledInEpoch(common.FixRelayedBaseCostFlag, epoch) {
```
Comment (sasurobert) [inline outport/process/transactionsfee/transactionsFeeProcessor.go:274]: do you need check.IfNilReflect ? don't you have simple check.IfNil ?

File: outport/process/transactionsfee/transactionsFeeProcessor.go [hunk 2]
```diff
@@ -157,6 +163,20 @@ func (tep *transactionsFeeProcessor) prepareTxWithResults(txHashHex string, txWi
 		}
 
 		if isSCRForSenderWithRefund(scr, txHashHex, txWithResults.GetTxHandler()) || isRefundForRelayed(scr, txWithResults.GetTxHandler()) {
+			if !check.IfNilReflect(userTx) && tep.enableEpochsHandler.IsFlagEnabledInEpoch(common.FixRelayedBaseCostFlag, epoch) {
```
Comment (sasurobert) [inline outport/process/transactionsfee/transactionsFeeProcessor.go:166]: do you need check.IfNilReflect ? don't you have simple check.IfNil ?

File: outport/process/transactionsfee/transactionsFeeProcessor.go (hunk 1 above)
Comment (sstanculeanu) [reply outport/process/transactionsfee/transactionsFeeProcessor.go:274]: updated

File: outport/process/transactionsfee/transactionsFeeProcessor.go (hunk 2 above)
Comment (sstanculeanu) [reply outport/process/transactionsfee/transactionsFeeProcessor.go:166]: updated

File: outport/process/transactionsfee/transactionsFeeProcessor.go [hunk 3]
```diff
@@ -157,6 +163,20 @@ func (tep *transactionsFeeProcessor) prepareTxWithResults(txHashHex string, txWi
 		}
 
 		if isSCRForSenderWithRefund(scr, txHashHex, txWithResults.GetTxHandler()) || isRefundForRelayed(scr, txWithResults.GetTxHandler()) {
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply outport/process/transactionsfee/transactionsFeeProcessor.go:165]: updated as discussed

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go [hunk 4]
```diff
@@ -132,32 +134,45 @@ func (gfp *gasUsedAndFeeProcessor) handleRelayedV2(args [][]byte, tx *transactio
...
 	}
 }
 
-func (gfp *gasUsedAndFeeProcessor) setGasUsedAndFeeBaseOnLogEvent(tx *transaction.ApiTransactionResult, hasRefund bool, event *transaction.Events) {
+func (gfp *gasUsedAndFeeProcessor) setGasUsedAndFeeBaseOnLogEvent(tx *transaction.ApiTransactionResult, userTx *transaction.ApiTransactionResult, hasRefund bool, event *transaction.Events) {
 	if core.WriteLogIdentifier == event.Identifier && !hasRefund {
-		gasUsed, fee := gfp.feeComputer.ComputeGasUsedAndFeeBasedOnRefundValue(tx, big.NewInt(0))
-		tx.GasUsed = gasUsed
-		tx.Fee = fee.String()
+		if !check.IfNilReflect(userTx) && gfp.enableEpochsHandler.IsFlagEnabledInEpoch(common.FixRelayedBaseCostFlag, tx.Epoch) {
```
Comment (sasurobert) [inline node/external/transactionAPI/gasUsedAndFeeProcessor.go:162]: check.IfNil

File: node/external/transactionAPI/gasUsedAndFeeProcessor.go [hunk 5]
```diff
@@ -166,14 +181,29 @@ func (gfp *gasUsedAndFeeProcessor) setGasUsedAndFeeBaseOnLogEvent(tx *transactio
 	}
 }
 
-func (gfp *gasUsedAndFeeProcessor) setGasUsedAndFeeBaseOnRefundValue(tx *transaction.ApiTransactionResult, refund *big.Int) {
+func (gfp *gasUsedAndFeeProcessor) setGasUsedAndFeeBaseOnRefundValue(
+	tx *transaction.ApiTransactionResult,
+	userTx *transaction.ApiTransactionResult,
+	refund *big.Int,
+) {
+	if !check.IfNilReflect(userTx) && gfp.enableEpochsHandler.IsFlagEnabledInEpoch(common.FixRelayedBaseCostFlag, tx.Epoch) {
```
Comment (sasurobert) [inline node/external/transactionAPI/gasUsedAndFeeProcessor.go:189]: check.IfNil

**Review by sasurobert — APPROVED** [review]

**Review by miiu96 — APPROVED** [review]

File: outport/process/transactionsfee/transactionsFeeProcessor.go [hunk 6]
```diff
@@ -249,6 +271,19 @@ func (tep *transactionsFeeProcessor) prepareTxWithResultsBasedOnLogs(
 
 	for _, event := range txWithResults.log.GetLogEvents() {
 		if core.WriteLogIdentifier == string(event.GetIdentifier()) && !hasRefund {
+			if !check.IfNil(userTx) && tep.enableEpochsHandler.IsFlagEnabledInEpoch(common.FixRelayedBaseCostFlag, epoch) {
```
Comment (AdoAdoAdo) [inline outport/process/transactionsfee/transactionsFeeProcessor.go:274]: same here the two code blocks inside prepareTxWithResultsBasedOnLogs and in prepareScrsNoTx are similar

Comment (sstanculeanu) [reply outport/process/transactionsfee/transactionsFeeProcessor.go:274]: done in https://github.com/multiversx/mx-chain-go/pull/6546

**Pull request merged**